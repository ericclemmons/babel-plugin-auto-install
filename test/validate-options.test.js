const InstallPlugin = require('../src/plugin');

describe('validation', () => {
  const cases = {
    dev: {
      success: [true, false, () => {}],
      failure: ['foo', 10],
    },
    npm: {
      success: [true, false],
      failure: ['bar', 10],
    },
    peerDependencies: {
      success: [true, false],
      failure: ['bar', 10],
    },
    prompt: {
      success: [true, false],
      failure: ['bar', 10],
    },
    quiet: {
      success: [true, false],
      failure: ['bar', 10],
    },
    yarn: {
      success: [true, false],
      failure: ['bar', 10],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === 'object' && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === 'success' ? 'successfully validate' : 'throw an error on'
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      let error;

      try {
        // eslint-disable-next-line no-new
        new InstallPlugin({ [key]: value });
      } catch (errorFromPlugin) {
        if (errorFromPlugin.name !== 'ValidationError') {
          throw errorFromPlugin;
        }

        error = errorFromPlugin;
      } finally {
        if (type === 'success') {
          expect(error).toBeUndefined();
        } else if (type === 'failure') {
          expect(() => {
            throw error;
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(cases)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
