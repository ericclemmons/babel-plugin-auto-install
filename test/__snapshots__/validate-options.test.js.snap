// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation should throw an error on the "dependencies" option with "{"peer":"bar"}" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.dependencies.peer should be a boolean.
   -> Install missing peer dependencies"
`;

exports[`validation should throw an error on the "dependencies" option with "{"peer":10}" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.dependencies.peer should be a boolean.
   -> Install missing peer dependencies"
`;

exports[`validation should throw an error on the "npm" option with "10" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.npm should be a boolean.
   -> Use NPM to install dependencies."
`;

exports[`validation should throw an error on the "npm" option with "bar" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.npm should be a boolean.
   -> Use NPM to install dependencies."
`;

exports[`validation should throw an error on the "packageManager" option with "{"type":"foo"}" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.packageManager.type should be one of these:
   \\"npm\\" | \\"pnpm\\" | \\"yarn\\""
`;

exports[`validation should throw an error on the "packageManager" option with "{"type":"npm","options":{"dev":"foo"}}" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.packageManager.options.dev should be a boolean.
   -> Install as development dependencies"
`;

exports[`validation should throw an error on the "packageManager" option with "foo" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.packageManager should be one of these:
   \\"npm\\" | \\"pnpm\\" | \\"yarn\\" | object { type?, options? } | function
   -> Package manager to use for installing dependencies.
   Details:
    * options.packageManager should be one of these:
      \\"npm\\" | \\"pnpm\\" | \\"yarn\\"
    * options.packageManager should be an object:
      object { type?, options? }
      -> package manager related options.
    * options.packageManager should be an instance of function."
`;

exports[`validation should throw an error on the "prompt" option with "10" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.prompt should be a boolean.
   -> Show a prompt to confirm installation."
`;

exports[`validation should throw an error on the "prompt" option with "bar" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.prompt should be a boolean.
   -> Show a prompt to confirm installation."
`;

exports[`validation should throw an error on the "quiet" option with "10" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.quiet should be a boolean.
   ->  Reduce amount of console logging"
`;

exports[`validation should throw an error on the "quiet" option with "bar" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.quiet should be a boolean.
   ->  Reduce amount of console logging"
`;

exports[`validation should throw an error on the "yarn" option with "10" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.yarn should be a boolean.
   -> Use yarn to install dependencies."
`;

exports[`validation should throw an error on the "yarn" option with "bar" value 1`] = `
"Invalid options object. Install Plugin has been initialized using an options object that does not match the API schema.
 - options.yarn should be a boolean.
   -> Use yarn to install dependencies."
`;
